Prevent potential memory leaks in CoreML backend

jmdetloff opened the pull request:
Differential Revision: D37266033

mcr229 reviewed (approved):
reviewed in phabricator D37266033

jmdetloff committed:
Prevent potential memory leaks in CoreML backend (#79928)

Summary:
Separate objc and c++ code as much as possible, then turn off ARC in PTMCoreMLBackend. This is necessary because we throw exceptions from within compile, and throwing exceptions in arc code leaks memory.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/79928

Test Plan:
To reproduce leak:
- Add an override to https://www.internalfb.com/intern/qe2/ig_ios_person_segmentation_universe to opt into the coreml segmentation model
- Drop a breakpoint on dealloc for PTMCoreMLExecutor so you can observe that it is not hit.
- Force an exception to throw during PTMCoreMLBackend compile.
- Launch IG camera and apply an effect that uses segmentation, such as green screen
- Observe that dealloc is not hit, the PTMCoreMLExecutor is leaked, along with the model.

To demo the fix:
- Perform the same steps as above and confirm that now dealloc is hit, PTMCoreMLExecutor is not leaked.

https://pxl.cl/26rk1

Reviewed By: kimishpatel, mcr229

Differential Revision: D37266033

Pulled By: jmdetloff

fbshipit-source-id: 4da83585f4e21c0e0ae602e5b18b962285983beb
